### DISCUSSION TITLE
Using Filament::getTenant() for Spark's billable resolver

### DISCUSSION THREAD

**Discussion by kurucu** [filamentphp/filament#8364] [Help]
Hi all, before I try and create a repo that demonstrates a bug I think I've found; can you sense check the following?

I set up multi tenancy with a Team tenant.
I set up Spark (Stripe) to use the Team as a billable.

In the SparkServiceProvider, I do the following:

```php
Spark::billable(Team::class)->resolve(function (Request $request) {
    return Filament::getTenant(); // This returns null
});
```

And for some reason, this returns `null`, which means after checking out and returning to Spark, the redirect goes to a 404 rather than the billing portal for that Team.

Should it be returning null? Should I be doing this another way?

Note: I only have one team, as I'm testing. Does this make any difference, as I've never had to "select" the Team? It seems the tenant is set in a middleware, so shouldn't matter?

**Comment by danharrin** [comment]
From which route is this function running? If its a non-Filament one (Spark-provided), the middleware wont run

**Reply by kurucu** [reply]
🤦‍♂️ Yep, I assumed it cached the last tenant somehow.

Is overriding the setTenant and getTenant to cache the last tenant for the session mad?

And - how do you see this working for Spark? I only ask because otherwise this has all been made so easy by Filament, and assume it should be part of the "normal" flow - but as it stands, no tenant-based checkouts can be completed.

Thinking about why it worked before Filament, Jetstream must have cached or put the last team in the session.

**Reply by kurucu** [reply]
So here's what I've done, which I think works.

I've created a tenant middleware and registered it with the panel, as follows:

```php
public function handle(Request $request, Closure $next)
{
    $request->session()->put('tenant_id', Filament::getTenant()->getKey());

    return $next($request);
}
```

```php
->tenantMiddleware([
  RememberTenantMiddleware::class,
])
```

This sets the last visited tenant id in the session. I did that, rather than a cache or user table, as it's feasible that the user has multiple sessions going.

Then, I added `HasDefaultTenant` and implemented as so for the User. This gets the last visited tenant, or otherwise just an arbitrary tenant (maybe I should see what the default is, and do that instead).

```php
public function getDefaultTenant(Panel $panel): ?Model {
    return $this->teams()->whereTeamId(session()->get('tenant_id'))->first() ?? $this->teams()->first();
}
```

And finally, `SparkServiceProvider` now resolves as follows. Again, the tenant panel doesn't vary here, so it's hard coded. Others might need something dynamic.

```php
Spark::billable(Team::class)->resolve(function (Request $request) {
    return $request->user()->getDefaultTenant(Filament::getPanel('team'));
});
```

**Reply by kurucu** [reply]
If this is a sensible solution, I'll make a Filament community item for it.

I guess I could have missed out the middle step and just put that logic in the resolve callback. This feels better somehow.

**Reply by danharrin** [reply]
Caching is a good idea, I'd love an article about it

**Reply by scottzirkel** [reply]
I've also set it up in the past where the active tenant id is stored on the user column so no matter what device they log in with, it always loads the same tenant.
